### PR TITLE
The javadocs for the templates are completed.The operation descriptions are added.

### PR DESCRIPTION
**Pull request by cetinkilic**

### CODE REVIEW

File: hazelcast-client-new/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java [hunk 1]
```diff
@@ -13,7 +13,7 @@
 import com.hazelcast.client.connection.ClientConnectionManager;
 import com.hazelcast.client.impl.client.DistributedObjectInfo;
 import com.hazelcast.client.impl.protocol.ClientMessage;
-import com.hazelcast.client.impl.protocol.codec.ClientGetDistributedObjectsCodec;
+import com.hazelcast.client.impl.protocol.codec.ClientGetDistributedObjectCodec;
```
Comment (sancar) [inline hazelcast-client-new/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: This looks like changed when merging. 
Name with plural is more appropriate since it returns a collection of distributed objects.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ClientMessageTemplate.java [hunk 2]
```diff
@@ -31,6 +31,9 @@
      * @param password Password for the user.
      * @param uuid Unique string identifying the connected client uniquely. This string is generated by the owner member server
      *             on initial connection. When the client connects to a non-owner member it sets this field on the request.
+     *             A UUID represents a 128-bit value. This string is a randomly generated uid based on
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ClientMessageTemplate.java]: This also looks like added back when merging. It is better to remove detailed explanation related to UUID

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ClientMessageTemplate.java [hunk 3]
```diff
@@ -38,12 +41,15 @@
...
     @Request(id = 2, retryable = true, response = ResponseMessageConst.AUTHENTICATION)
     Object authentication(String username, String password, @Nullable String uuid,
-                        @Nullable String ownerUuid, boolean isOwnerConnection, String clientType);
+                          @Nullable String ownerUuid, boolean isOwnerConnection, String clientType);
     /**
      *
      * @param credentials Secret byte array for authentication.
      * @param uuid Unique string identifying the connected client uniquely. This string is generated by the owner member server
      *             on initial connection. When the client connects to a non-owner member it sets this field on the request.
+     *             A UUID represents a 128-bit value. This string is a randomly generated uid based on
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ClientMessageTemplate.java]: Same here. Better to remove it.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/CacheCodecTemplate.java [hunk 4]
```diff
@@ -174,16 +197,19 @@
...
+     * their appropriate CacheEntryReadListeners notified and b). removed will have their appropriate
+     * CacheEntryRemoveListeners notified. java.util.Iterator#next() may return null if the entry is no longer present,
+     * has expired or has been evicted.
      *
      * @param name Name of the cache.
      * @param partitionId The partition id which owns this cache store.
      * @param tableIndex The slot number (or index) to start the iterator
      * @param batch The number of items to be batched
-     * @return last index processed and list of data
+     * @return Serialised com.hazelcast.cache.impl.CacheKeyIteratorResult object.
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/CacheCodecTemplate.java]: Also changed while merging. Old description suits better here.

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java [hunk 5]
```diff
@@ -155,6 +155,21 @@ public static String getReturnDescription(String commentString) {
         return result;
     }
 
+    public static String getOperationDescription(String commentString){
+        String result="";
+        int nextTagIndex = commentString.indexOf("@");
+        if (nextTagIndex >= 0) {
+            result = commentString.substring(0,nextTagIndex);
+
+        result = result.trim();
```
Comment (ihsandemir) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java]: indentation.

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodecModel.java [hunk 6]
```diff
@@ -20,6 +20,7 @@
 import com.hazelcast.annotation.GenerateCodec;
 import com.hazelcast.annotation.Nullable;
 import com.hazelcast.annotation.Request;
+import com.sun.org.apache.bcel.internal.classfile.Code;
```
Comment (ihsandemir) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodecModel.java]: Why do we need this import? New dependency?

File: hazelcast-code-generator/src/main/resources/codec-template-md.ftl [hunk 7]
```diff
@@ -259,7 +262,10 @@ Header only event message, no message body exist.
 </#list>
 
 <#if key == "com.hazelcast.client.impl.protocol.template.EnterpriseMapCodecTemplate">
-<b>Note:</b> All operation defined for the Map Object can also be executed against the EnterpriseMap Object.
+
+<b>Note:</b> All operation defined for the Map Object can also be executed against the EnterpriseMap Object as well.
+ The javadocs for the templates are completed. The event parameter descriptions are added.
```
Comment (ihsandemir) [inline hazelcast-code-generator/src/main/resources/codec-template-md.ftl]: Remove the added sentence. It may be added by mistake.

File: hazelcast-code-generator/src/test/java/com/hazelcast/codegenerator/GeneratorTest.java [hunk 8]
```diff
@@ -45,7 +50,26 @@ public GeneratorTest() {
         }
 
     }
+    @Test
+    public void operationDescriptionTest(){
```
Comment (ihsandemir) [inline hazelcast-code-generator/src/test/java/com/hazelcast/codegenerator/GeneratorTest.java]: The test does not test anything but just does a println.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ClientMessageTemplate.java (hunk 3 above)
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ClientMessageTemplate.java]: This was removed at previous commits.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ClientMessageTemplate.java [hunk 9]
```diff
@@ -164,4 +170,4 @@ Object authenticationCustom(Data credentials, @Nullable String uuid,
 
     @Request(id = 15, retryable = true, response = ResponseMessageConst.VOID)
     void ping();
-}
+}
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ClientMessageTemplate.java]: Re-add the new line.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ConditionCodecTemplate.java [hunk 10]
```diff
@@ -44,15 +61,23 @@
     void beforeAwait(String name, long threadId, String lockName);
 
     /**
+     * If any threads are waiting on this condition then one is selected for waking up.That thread must then re-acquire
+     * the lock before returning from await. An implementation may (and typically does) require that the
+     * current thread hold the lock associated with this Condition when this method is called. Implementations must
+     * document this precondition and any actions taken if the lock is not held. Typically, an exception such as
+     * IllegalMonitorStateException will be thrown.
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ConditionCodecTemplate.java]: Need to put the error code for the protocol. Exception name is not meaningful.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ClientMessageTemplate.java [hunk 11]
```diff
@@ -145,7 +151,7 @@ Object authenticationCustom(Data credentials, @Nullable String uuid,
      * @return An array of distributed object info in the cluster.
      */
     @Request(id = 12, retryable = false, response = ResponseMessageConst.SET_DISTRIBUTED_OBJECT)
-    Object getDistributedObjects();
+    Object getDistributedObject();
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ClientMessageTemplate.java]: should be getDistributedObjects()

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/CountdownLatchCodecTemplate.java [hunk 12]
```diff
@@ -33,27 +43,33 @@
...
-     * @param count The number of times countDown must be invoked before threads can pass through await
+     * @param name Name of the CountdownLatch
+     * @param count The number of times  countDown must be invoked before threads can pass through await
      * @return True if the new count was set, false if the current count is not zero.
      */
     @Request(id = 4, retryable = false, response = ResponseMessageConst.BOOLEAN)
     Object trySetCount(String name, int count);
 
-}
+}
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/CountdownLatchCodecTemplate.java]: Re-add the new line.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/EnterpriseMapCodecTemplate.java [hunk 13]
```diff
@@ -40,7 +40,7 @@
      */
     @Request(id = 1, retryable = true, response = ResponseMessageConst.SET_ENTRY)
     Object publisherCreateWithValue(String mapName, String cacheName, Data predicate, int batchSize, int bufferSize,
-                                  long delaySeconds, boolean populate, boolean coalesce);
+                                    long delaySeconds, boolean populate, boolean coalesce);
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/EnterpriseMapCodecTemplate.java]: No difference, just revert the change.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/EnterpriseMapCodecTemplate.java [hunk 14]
```diff
@@ -57,7 +57,7 @@ Object publisherCreateWithValue(String mapName, String cacheName, Data predicate
      */
     @Request(id = 2, retryable = true, response = ResponseMessageConst.SET_DATA)
     Object publisherCreate(String mapName, String cacheName, Data predicate, int batchSize, int bufferSize, long delaySeconds,
-                         boolean populate, boolean coalesce);
+                           boolean populate, boolean coalesce);
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/EnterpriseMapCodecTemplate.java]: Revert the change.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/EnterpriseMapCodecTemplate.java [hunk 15]
```diff
@@ -96,4 +96,4 @@ Object publisherCreate(String mapName, String cacheName, Data predicate, int bat
     @Request(id = 6, retryable = false, response = ResponseMessageConst.BOOLEAN)
     Object destroyCache(String mapName, String cacheName);
 
-}
+}
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/EnterpriseMapCodecTemplate.java]: revert

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/EnterpriseMapCodecTemplate.java [hunk 16]
```diff
@@ -74,7 +74,7 @@ Object publisherCreate(String mapName, String cacheName, Data predicate, int bat
      * @return Registration id for the listener.
      */
     @Request(id = 4, retryable = true, response = ResponseMessageConst.STRING,
-             event = {EventMessageConst.EVENT_QUERYCACHESINGLE, EventMessageConst.EVENT_QUERYCACHEBATCH})
+            event = {EventMessageConst.EVENT_QUERYCACHESINGLE, EventMessageConst.EVENT_QUERYCACHEBATCH})
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/EnterpriseMapCodecTemplate.java]: revert

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/EventResponseTemplate.java [hunk 17]
```diff
@@ -188,5 +190,4 @@ void Entry(@Nullable Data key, @Nullable Data value, @Nullable Data oldValue, @N
 
     @EventResponse(EventMessageConst.EVENT_CACHEPARTITIONLOST)
     void CachePartitionLost(int partitionId, String uuid);
-
-}
+}
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/EventResponseTemplate.java]: revert

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/EventResponseTemplate.java [hunk 18]
```diff
@@ -21,6 +22,7 @@
 import com.hazelcast.annotation.Nullable;
 import com.hazelcast.cache.impl.CacheEventData;
 import com.hazelcast.client.impl.protocol.EventMessageConst;
+import com.hazelcast.cluster.client.MemberAttributeChange;
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/EventResponseTemplate.java]: revert

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/TransactionalMultiMapCodecTemplate.java [hunk 19]
```diff
@@ -26,9 +26,10 @@
         name = "TransactionalMultiMap", ns = "Hazelcast.Client.Protocol.TransactionalMultiMap")
 public interface TransactionalMultiMapCodecTemplate {
     /**
+     * Stores a key-value pair in the multimap.
      *
      * @param name Name of the Transactional Multi Map
-     * @param txnId ID of the transaction
+     * @param txnId ID of the this transaction
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/TransactionalMultiMapCodecTemplate.java]: merge issue? The world 'this' should be removed on all similar methods

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ListCodecTemplate.java [hunk 20]
```diff
@@ -116,24 +140,28 @@
...
     /**
+     * Adds an item listener for this collection. Listener will be notified for all collection add/remove events.
      *
      * @param name Name of the List
      * @param includeValue Set to true if you want the event to contain the value.
-     * @return Registration id for the listener.
+     * @return Registration id for the listener
      */
-    @Request(id = 11, retryable = true, response = ResponseMessageConst.STRING, event = {EventMessageConst.EVENT_ITEM})
+    @Request(id = 11, retryable = true, response = ResponseMessageConst.STRING
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ListCodecTemplate.java]: revert

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ListCodecTemplate.java [hunk 21]
```diff
@@ -142,17 +170,24 @@
...
-     * @param index index at which to insert the first element from the
-     *              specified collection
-     * @param valueList The list of value to insert into the list.
-     * @return True if this list changed as a result of the call, false otherwise.
+     * @param index index at which to insert the first element from the specified collection
+     * @param valueList The list of value to insert into the list
+     * @return True if this list changed as a result of the call, false otherwise
+     *
      */
+    //LIST PARAMS
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ListCodecTemplate.java]: Delete this line

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ListCodecTemplate.java [hunk 22]
```diff
@@ -190,6 +230,9 @@
     Object removeWithIndex(String name, int index);
 
     /**
+     * Returns the index of the last occurrence of the specified element in this list, or -1 if this list does not
+     * contain the element. More formally, returns the highest index i such that
+     * (value==null ? get(i)==null : value.equals(get(i))), or -1 if there is no such index.
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ListCodecTemplate.java]: Can we explain this without get(i) usage, more in terms of the protocol operations?

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/LockCodecTemplate.java [hunk 23]
```diff
@@ -91,4 +105,4 @@
     @Request(id = 8, retryable = false, response = ResponseMessageConst.BOOLEAN)
     Object tryLock(String name, long threadId, long lease, long timeout);
 
-}
+}
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/LockCodecTemplate.java]: revert

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/MapCodecTemplate.java [hunk 24]
```diff
@@ -20,14 +20,22 @@
 import com.hazelcast.annotation.Request;
 import com.hazelcast.client.impl.protocol.EventMessageConst;
 import com.hazelcast.client.impl.protocol.ResponseMessageConst;
+import com.hazelcast.instance.GroupProperties;
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/MapCodecTemplate.java]: why added these imports?

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/MapCodecTemplate.java [hunk 25]
```diff
@@ -453,6 +570,10 @@
     Object valuesWithPredicate(String name, Data predicate);
 
     /**
+     * Queries the map based on the specified predicate and returns the matching entries.Specified predicate
+     * runs on all members in parallel.The collection is NOT backed by the map, so changes to the map are NOT reflected
+     * in the collection, and vice-versa.This method is always executed by a distributed query so it may throw a
+     * QueryResultSizeExceededException if GroupProperties#PROP_QUERY_RESULT_SIZE_LIMIT is configured.
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/MapCodecTemplate.java]: use error code

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/MapCodecTemplate.java [hunk 26]
```diff
@@ -572,6 +723,10 @@
     Object keySetWithPagingPredicate(String name, Data predicate);
 
     /**
+     * Queries the map based on the specified predicate and returns the values of matching entries. Specified predicate
+     * runs on all members in parallel. The collection is NOT backed by the map, so changes to the map are NOT reflected
+     * in the collection, and vice-versa. This method is always executed by a distributed query so it may throw a
+     * QueryResultSizeExceededException if GroupProperties#PROP_QUERY_RESULT_SIZE_LIMIT is configured.
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/MapCodecTemplate.java]: error code

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/MapCodecTemplate.java [hunk 27]
```diff
@@ -590,4 +745,4 @@
     @Request(id = 58, retryable = false, response = ResponseMessageConst.SET_ENTRY)
     Object entriesWithPagingPredicate(String name, Data predicate);
 
-}
+}
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/MapCodecTemplate.java]: revert

**Comment by cetinkilic** [comment]
@cangencer @ihsandemir @sancar thanks for the commets. i have updated the PR

**Comment by sancar** [comment]
:+1: Looks good

**Comment by cangencer** [comment]
:+1: looks much better!

**Comment by ihsandemir** [comment]
:+1:
